src/features/device-state/routes/state-patch-v3.ts

```diff
@@ -0,0 +1,384 @@
+							releases,
+						)) {
+							appReleaseUuids[appUuid].add(releaseUuid);
+							if (services != null) {
+								for (const [, service] of Object.entries(services)) {
+									imageLocations.push(service.image);
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+
+		let devices: Array<
+			Pick<Device, 'id' | 'uuid'> & {
+				belongs_to__application: Array<Pick<Application, 'uuid'>>;
+			}
+		>;
+		let images: Array<Pick<Image, 'id' | 'is_stored_at__image_location'>>;
+		const appReleases: {
+			[appUuid: string]: Array<Pick<Release, 'id' | 'commit'>>;
+		} = {};
+		await sbvrUtils.db.readTransaction(async (tx) => {
+			const resinApiTx = api.resin.clone({
+				passthrough: { req, custom, tx },
+			});
+			devices = (await resinApiTx.get({
+				resource: 'device',
+				options: {
+					$select: ['id', 'uuid', 'belongs_to__application'],
+					$filter: {
+						uuid: { $in: uuids },
+					},
+					$expand: {
+						belongs_to__application: {
+							$select: 'uuid',
+						},
+					},
+				},
+			})) as Array<
+				Pick<Device, 'id' | 'uuid'> & {
+					belongs_to__application: Array<Pick<Application, 'uuid'>>;
+				}
+			>;
+			if (devices.length !== uuids.length) {
+				throw new UnauthorizedError();
+			}
+
+			if (imageLocations.length > 0) {
+				const imgLocationFilter = imageLocations.map((imgLocation) => {
+					const [location, contentHash] = imgLocation.split('@');
+					const filter: Filter = { is_stored_at__image_location: location };
+					if (contentHash) {
+						filter.content_hash = contentHash;
+					}
+					return filter;
+				});
+				images = (await resinApiTx.get({
+					resource: 'image',
+					options: {
+						$select: ['id', 'is_stored_at__image_location'],
+						$filter:
+							imgLocationFilter.length === 1
+								? imgLocationFilter[0]
+								: imgLocationFilter,
+					},
+				})) as Array<Pick<Image, 'id' | 'is_stored_at__image_location'>>;
+				if (imageLocations.length !== images.length) {
+					throw new UnauthorizedError();
+				}
+			}
+
+			for (const [appUuid, releaseUuids] of Object.entries(appReleaseUuids)) {
+				appReleases[appUuid] =
+					releaseUuids.size === 0
+						? []
+						: ((await resinApiTx.get({
+								resource: 'release',
+								options: {
+									$select: ['id', 'commit'],
+									$filter: {
+										commit: { $in: Array.from(releaseUuids) },
+										belongs_to__application: {
+											$any: {
+												$alias: 'a',
+												$expr: {
+													a: { uuid: appUuid },
+												},
+											},
+										},
+									},
+								},
+						  })) as Array<Pick<Release, 'id' | 'commit'>>);
+				if (appReleases[appUuid].length !== releaseUuids.size) {
+					throw new UnauthorizedError();
+				}
+			}
+		});
+
+		await sbvrUtils.db.transaction(async (tx) => {
+			const resinApiTx = api.resin.clone({
+				passthrough: { req, custom, tx },
+			});
+
+			const waitPromises: Array<PromiseLike<any>> = [];
```

Comment:
Would it be of a benefit if this was 
```suggestion
			const updateFns: Array<(resinApiTx: PinejsClient) => Promise<void>> = [];
```
like in in v2?

Reply:
Right, yeah I should probably update this to match those v2 changes

Reply:
Done